### ISSUE TITLE
No pressure sensitivity for dots

### ISSUE THREAD

**Issue by JLTastet** [xournalpp/xournalpp#1534]
**Affects versions :**
 - OS: Linux (Fedora 30)
 - Gnome-Wayland 3.32.2
 - Which version of libgtk do you use: not sure
 - Version of Xournal++: 1.0.10, version `0+git.26c7205f` from snapcraft.io

**Describe the bug**
When trying to draw small dots using the stylus, the resulting dots always have the same size (which is too big), hence ignoring the stylus pressure, as can be seen on the screenshot below. They also seem to appear with a small delay, like when using the shape recogniser.
I cannot figure out if this is a feature, an issue with my tablet, or some X++ bug. It affect the original Xournal too.

**To Reproduce**
1. Quickly tap the tablet/screen with the stylus to produce a dot.
2. For reference, repeat step 1, but this time move the cursor very slightly while applying pressure.

**Expected behavior**
Step 1 generates a small dot, with a size similar to normal strokes and with pressure sensitivity. The result should look similar to that of step 2.

**Actual behaviour**
Step 1 generates a large, fixed size dot like in the screenshot below. 

**Screenshot of Problem**
![Screenshot from 2019-10-11 18-26-28](https://user-images.githubusercontent.com/8004066/66668831-6b1e4800-ec56-11e9-8b09-8f32224e23fd.png)

**Attachments**

* Zipped `.xopp` file: [Dots.zip](https://github.com/xournalpp/xournalpp/files/3718748/Dots.zip)

**Comment by Febbe** [comment]
I am not Shure, but dots should be represented by a single point in a stroke. And the last point in a stroke has in fact no pressure. Maybe you hit that cornercase. We should consider to use the average pressure between 2 points and store in each point the pressure.

**Comment by JLTastet** [comment]
Do you know if there is a technical reason why the last point has no pressure? Or did X++ just inherit this behaviour from Xournal?
Also, do I understand correctly that I just see this behaviour because for some reason the strokes happen to contain only one point, or are points somehow treated as a special case?

**Comment by alfazaz** [comment]
Same bug for me... and it's really annoying on writing text with letters like "i"... :-(

**Comment by ViniciusGiroto** [comment]
I was going to open a bug report about that. Can we expect a fix any time soon?

**Comment by rolandlo** [comment]
In version 1.1.0 you can adjust the pressure multiplier in the Stylus panel of menu `Edit > Preferences`.

**Comment by JLTastet** [comment]
> In version 1.1.0 you can adjust the pressure multiplier in the Stylus panel of menu `Edit > Preferences`.

Indeed this partially solves the issue, since it allows selecting a thinner stroke width, resulting in smaller dots (just tested on `1.1.0-9d1277de`).

The size of the dots is still not pressure sensitive, though, and whether "pressing the pen on the screen" produces a dot is still quite hit-or-miss, so this feature can still be a bit annoying.

**Comment by bhennion** [comment]
The source of the issue is [this line](https://github.com/xournalpp/xournalpp/blob/3fe13126e893f88f461cf48cf891f74654b73b63/src/control/tools/StrokeHandler.cpp#L227)
It first appeared in c97f1c2 back in 2011, probably to mimick Xournal.

I imagine the reason for that is that some devices don't give a pressure value (or more annoyingly give 0 or another very small value) for lonely events like that. I'll try to think of a better solution.

**Comment by Febbe** [comment]
I think the easiest solution would be, to add a second point to the stroke, when it's a dot. 

The cleaner solution, would be to overthrow the old behavior and add pressure to each point. But this can have problems on ser/des.

**Comment by bhennion** [comment]
> I think the easiest solution would be, to add a second point to the stroke, when it's a dot.

This is already what happens on [this line](https://github.com/xournalpp/xournalpp/blob/3fe13126e893f88f461cf48cf891f74654b73b63/src/control/tools/StrokeHandler.cpp#L224).

> The cleaner solution, would be to overthrow the old behavior and add pressure to each point. But this can have problems on ser/des.

This is already a concern of mine in #2729. We should definitely discuss it somewhere, but it's off topic here, IMO.

The question for the bug at hand is how to get a meaningful pressure value on all devices for (seamingly) single-event strokes. I have `libinput` recordings of 3 different devices and they all have different behaviours when it comes to the pressure value when receiving a `BTN_TOUCH = 1` event (when the stylus touches the tablet/screen). For the first, the pressure value is about 1% of the average value, for the second, it's ~40% and for the third it's ~70% (those values are somewhat reproducible).

Maybe this kind of single-event-strokes are an artefact of the test performed [here](https://github.com/xournalpp/xournalpp/blob/3fe13126e893f88f461cf48cf891f74654b73b63/src/control/tools/StrokeHandler.cpp#L78-L80). Pressure increase could/should be recorded even if the tip did not move.
Here is an example of a sequence of `libinput` events that resulted in a single fat dot in `Xournal++`:
```
Event: time 1629831144.177751, -------------- SYN_REPORT ------------
Event: time 1629831144.180433, type 3 (EV_ABS), code 0 (ABS_X), value 5397
Event: time 1629831144.180433, type 3 (EV_ABS), code 1 (ABS_Y), value 15193
Event: time 1629831144.180433, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 1121
Event: time 1629831144.180433, type 1 (EV_KEY), code 330 (BTN_TOUCH), value 1
Event: time 1629831144.180433, -------------- SYN_REPORT ------------
Event: time 1629831144.186831, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 1812
Event: time 1629831144.186831, -------------- SYN_REPORT ------------
Event: time 1629831144.189723, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 2092
Event: time 1629831144.189723, -------------- SYN_REPORT ------------
Event: time 1629831144.192549, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 2168
Event: time 1629831144.192549, -------------- SYN_REPORT ------------
Event: time 1629831144.195268, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 2397
Event: time 1629831144.195268, -------------- SYN_REPORT ------------
Event: time 1629831144.201686, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 2415
Event: time 1629831144.201686, -------------- SYN_REPORT ------------
Event: time 1629831144.204574, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 2409
Event: time 1629831144.204574, -------------- SYN_REPORT ------------
Event: time 1629831144.207421, type 3 (EV_ABS), code 1 (ABS_Y), value 15192
Event: time 1629831144.207421, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 2387
Event: time 1629831144.207421, -------------- SYN_REPORT ------------
Event: time 1629831144.210114, type 3 (EV_ABS), code 1 (ABS_Y), value 15191
Event: time 1629831144.210114, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 2356
Event: time 1629831144.210114, -------------- SYN_REPORT ------------
Event: time 1629831144.216604, type 3 (EV_ABS), code 1 (ABS_Y), value 15190
Event: time 1629831144.216604, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 2303
Event: time 1629831144.216604, -------------- SYN_REPORT ------------
Event: time 1629831144.219377, type 3 (EV_ABS), code 1 (ABS_Y), value 15189
Event: time 1629831144.219377, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 2236
Event: time 1629831144.219377, -------------- SYN_REPORT ------------
Event: time 1629831144.222279, type 3 (EV_ABS), code 1 (ABS_Y), value 15188
Event: time 1629831144.222279, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 2148
Event: time 1629831144.222279, -------------- SYN_REPORT ------------
Event: time 1629831144.224897, type 3 (EV_ABS), code 1 (ABS_Y), value 15187
Event: time 1629831144.224897, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 2034
Event: time 1629831144.224897, -------------- SYN_REPORT ------------
Event: time 1629831144.231345, type 3 (EV_ABS), code 1 (ABS_Y), value 15186
Event: time 1629831144.231345, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 1843
Event: time 1629831144.231345, -------------- SYN_REPORT ------------
Event: time 1629831144.234199, type 3 (EV_ABS), code 1 (ABS_Y), value 15185
Event: time 1629831144.234199, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 1579
Event: time 1629831144.234199, -------------- SYN_REPORT ------------
Event: time 1629831144.237103, type 3 (EV_ABS), code 1 (ABS_Y), value 15184
Event: time 1629831144.237103, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 1493
Event: time 1629831144.237103, -------------- SYN_REPORT ------------
Event: time 1629831144.239835, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 1277
Event: time 1629831144.239835, -------------- SYN_REPORT ------------
Event: time 1629831144.246247, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 991
Event: time 1629831144.246247, -------------- SYN_REPORT ------------
Event: time 1629831144.249008, type 3 (EV_ABS), code 24 (ABS_PRESSURE), value 0
Event: time 1629831144.249008, type 1 (EV_KEY), code 330 (BTN_TOUCH), value 0
Event: time 1629831144.249008, -------------- SYN_REPORT ------------
```
With such sequences, we should just get the highest recorded pressure value and use it. I don't know that every device behaves this way though. I'll write up a PR for people to test on their devices.

**Comment by bhennion** [comment]
Fixed via #3344 via #3599
